### PR TITLE
Keyboard shortcuts inhibit

### PR DESCRIPTION
**Pull request by michaelweiser**
Adding support for the keyboard shortcuts inhibit protocol allows remote
desktop and virtualisation software to receive all keyboard input in
order to pass it through to their clients so users can fully interact
the their remote/virtual session. The software usually provides its own
key combination to release its "grab" to all keyboard input. The
inhibitor can be deactivated by the user by removing focus from the
surface using another input device such as the pointer.

Use support for the procotol in wlroots (swaywm/wlroots#2026) to add support to sway. Extend
the input manager with handlers for inhibitor creation and destruction
and appropriate bookkeeping. Attach the inhibitors to the seats they
apply to to avoid having to search the list of all currently existing
inhibitors on every keystroke and passing the inhibitor manager around.
Add a helper function to retrieve the inhibitor applying to the
currently focused surface of a seat, if one exists.

Extend bindsym with a flag for bindings that should be processed even if
an inhibitor is active. Conversely this disables all normal shortcuts if
an inhibitor is found for the currently focused surface in
keyboard::handle_key_event() since they don't have that flag set. Use
above helper function to determine if an inhibitor exists for the
surface that would eventually receive input.

Also, add a command to influence keyboard shortcuts inhibitors. In its current
form it can be used to activate, deactivate or toggle an existing
inhibitor on the surface currently receiving input. This can be used to
define an escape shortcut such as:
```
bindsym --inhibited $mod+Ctrl+Shift+q shortcuts_inhibitor deactivate
```

### CODE REVIEW

File: include/sway/config.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ enum binding_flags {
 	BINDING_TITLEBAR=16, // mouse only; trigger on container titlebar
 	BINDING_CODE=32,     // keyboard only; convert keysyms into keycodes
 	BINDING_RELOAD=62,   // switch only; (re)trigger binding on reload
```
Comment (michaelweiser) [inline include/sway/config.h]: OT: Shouldn't this be 64?

Comment (ddevault) [reply include/sway/config.h]: Yep.

Comment (michaelweiser) [reply include/sway/config.h]: Separate PR or drive-by fix in this one? Have no idea how to and no real inclination of going down the rabbithole of retesting it, TBH. :-/ Had a quick glance where it's used and seems trivial enough, though.

Comment (ddevault) [reply include/sway/config.h]: A second commit in this PR would be fine imo.

**Comment by michaelweiser** [comment]
I'm struggling with what to do with the compositor built-in bindings in `keyboard::handle_key_event()`. Initially I thought that switching VTs (which is all they're implemeting currently) is such a fundamental recovery technique that I shouldn't allow inhibitors to suppress them by default. So I made inhibitors not apply to them for now.

Instead I wanted to add an additional oneshot passthrough command similar to what some VM software does, i.e. have a keybinding that, if activated, will make sway ignore the next binding and instead pass it through to the client. That way you could press e.g. Ctrl+Alt+Space which would activate oneshot passthrough and then press Ctrl+Alt+F2 which would switch the VT in the remote or virtual session instead of locally.

I got a first version implemented but it was quite cumbersome and literally just made sway ignore the next binding, no matter how far in the future that may be and how many non-binding keystrokes came inbetween.

To be intuitive, oneshot passthrough should IMO also disable itself on any non-binding key combo that *was* indeed passed to the client. However, the way `keyboard::handle_key_event()` processes key events appears particularly ill suited for this kind of thing (or my brain is). Also, looking at the existing complexity my guess is that any additional complexity in that code path would better be avoided.

Long story short: Should I bother or could we instead have a default/recommended escape key binding for uncooperative clients/inhibitors?

File: include/sway/config.h (hunk 1 above)
Comment (michaelweiser) [reply include/sway/config.h]: Done. Thanks!

**Comment by ddevault** [comment]
Would it make sense to shelve that for a future patch?

**Comment by michaelweiser** [comment]
> Would it make sense to shelve that for a future patch?
Indeed. After some discussion and brainstorming (Thanks @wolf-z!) I came to realize that most virtualization and remote desktop software has virtual keyboards, shortcut sending menus or special key combos to work around exactly this issue. So I think the current implementation is in line with existing ones. I quickly retested with GNOME and virt-manager and found that it also does not inhibit Ctrl+Alt+Fx and virt-manager provides a menu to send them to the guest.

BTW: I also came across this article which mentions a warning dialog in GNOME that an application wants to inhibit the keyboard and that the escape key combo is Super+Escape: https://unix.stackexchange.com/questions/417670/virtual-machine-manager-wants-to-inhibit-shortcuts-again-and-again-on-waylan. I'm not saying we should have a warning dialog but I like the Super+Escape key combination.

File: sway/commands/shortcuts_inhibitor.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#include "sway/commands.h"
+#include "sway/input/seat.h"
+#include "sway/input/input-manager.h"
+
+struct cmd_results *cmd_shortcuts_inhibitor(int argc, char **argv) {
+	struct cmd_results *error = checkarg(argc, "shortcuts_inhibitor",
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/commands/shortcuts_inhibitor.c]: Done.

File: sway/commands/shortcuts_inhibitor.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+			EXPECTED_EQUAL_TO, 1);
+	if (error) {
+		return error;
+	}
+
+	struct sway_seat *seat = config->handler_context.seat;
+	struct sway_keyboard_shortcuts_inhibitor *sway_inhibitor =
+		keyboard_shortcuts_inhibitor_get_for_focused_surface(seat);
+	if (!sway_inhibitor) {
+		return cmd_results_new(CMD_SUCCESS, NULL);
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/commands/shortcuts_inhibitor.c]: Done.

File: sway/input/input-manager.c [hunk 4]
```diff
@@ -296,6 +296,48 @@ static void handle_inhibit_deactivate(struct wl_listener *listener, void *data)
...
+				"shortcuts inhibitor")) {
+		return;
+	}
+	sway_inhibitor->inhibitor = inhibitor;
+
+	sway_inhibitor->destroy.notify = handle_keyboard_shortcuts_inhibitor_destroy;
+	wl_signal_add(&inhibitor->events.destroy, &sway_inhibitor->destroy);
+
+	// attach inhibitor to the seat it applies to
+	struct sway_seat *seat = NULL;
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/input/input-manager.c]: Uncanny. :) Done.

File: sway/input/keyboard.c [hunk 5]
```diff
@@ -150,16 +150,18 @@ static bool update_shortcut_state(struct sway_shortcut_state *state,
...
 	for (int i = 0; i < bindings->length; ++i) {
 		struct sway_binding *binding = bindings->items[i];
 		bool binding_locked = (binding->flags & BINDING_LOCKED) != 0;
+		bool binding_inhibited = (binding->flags & BINDING_INHIBITED) != 0;
 		bool binding_release = binding->flags & BINDING_RELEASE;
 
 		if (modifiers ^ binding->modifiers ||
 				release != binding_release ||
 				locked > binding_locked ||
+				inhibited != binding_inhibited ||
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/input/keyboard.c]: I had this programmed out once already but then decided that I did not want --inhibited bindsyms to leak into normal modes. But I agree that I certainly also want to be able to mark my brightness and volume bindsyms as non-inhibited. So: Done. Thanks for making me rethink this.

File: sway/commands/shortcuts_inhibitor.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include "sway/input/input-manager.h"
+
+struct cmd_results *cmd_shortcuts_inhibitor(int argc, char **argv) {
+	struct cmd_results *error = checkarg(argc, "shortcuts_inhibitor",
+			EXPECTED_EQUAL_TO, 1);
+	if (error) {
+		return error;
+	}
+
+	struct sway_seat *seat = config->handler_context.seat;
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/commands/shortcuts_inhibitor.c]: Done. While at it I took inspiration from `pointer_constraint` and added a per-`seat_config` (and thus with `*` potentially global) enable/disable default. This actually allows the user to decide when to activate an inhibitor.

The inhibitor actions don't seem to fit into this all that well. I had to redefine the wildcard `seat *` to mean "current seat"  to keep semantics. What do you think of this?

What's also left is the question how the user could get some indication of an existing inhibitor on a window and its state. Maybe a border color change? I had thought of a criteria for that. What do you think and how would I go about it?

[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/commands/shortcuts_inhibitor.c]: I have the criteria coded up but it doesn't fully work because criteria are only executed once. So a config like:
```
for_window [shortcuts_inhibitor=active] title_format !+%title
for_window [shortcuts_inhibitor=inactive] title_format !-%title
for_window [shortcuts_inhibitor=absent] title_format %title
```
will happily upgrade the window title to '!+%title` as an inhibitor becomes active but not back down if it goes inactive or gets destroyed.

Now I'm thinking of executing a configurable command as the inhibitor state changes but I can't find any example for such a `command_to_exec_on_some_condition` hook-like functionality. What would the sway-native approach be here? Have a script listen to sway-ipc and send the `title_format` command?

[the remark this answers is not in the document]
Comment (RedSoxFan) [reply sway/commands/shortcuts_inhibitor.c]: > The inhibitor actions don't seem to fit into this all that well. I had to redefine the wildcard seat * to mean "current seat" to keep semantics. What do you think of this?

There is already `seat -` for the current seat. `seat *` should work fine for `enable|disable`. Either return a result with the status `CMD_INVALID` for `activate|deactivate` on the wildcard seat or activate/deactivate all shortcut inhibitors. I'm not really sure which is better.

> What's also left is the question how the user could get some indication of an existing inhibitor on a window and its state. Maybe a border color change? I had thought of a criteria for that. What do you think and how would I go about it?
> [..]
> What would the sway-native approach be here? Have a script listen to sway-ipc and send the title_format command?

Information for none of the inhibitor types is currently exposed via IPC. It could be added to get_tree response though. Any other change would be out of scope for this PR and would have to be discussed as a separate one.

File: sway/input/keyboard.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (RedSoxFan) [reply sway/input/keyboard.c]: Just to explain the reason why we have it this way, it is to remove duplication of bindings. You can always specify different actions (or no action with `nop`) for a more specific case.
```
# Both
bindsym $mod+Escape --inhibited exec notify-send "Both Normal and Inhbited"

# Just Normal
bindsym $mod+Escape exec notify-send "Only Normal Shortcuts"

# Just Inhibited (order doesn't matter)
bindsym $mod+Escape nop
bindsym $mod+Escape --inhibited notify-send "Only Inhibited"
```

File: sway/commands/seat/shortcuts_inhibitor.c [hunk 7]
```diff
@@ -0,0 +1,121 @@
+#include "sway/commands.h"
```
Comment (RedSoxFan) [inline sway/commands/seat/shortcuts_inhibitor.c]: Not a huge fan of this entire file. I'd recommend taking inspiration from https://github.com/swaywm/sway/blob/e1eda8a09c15780e7f6ab0b6a1ab7e811a3d8764/sway/commands/urgent.c

Something like:
```c
// Validate config->handler_context.seat_config is not NULL
if (strcmp(argv[0], "enable") == 0) {  // Possibly rename to `allow` for consistency
    sc->allow_shortcuts_inhibit = true;
} else if (strcmp(argv[0], "disable") == 0) {  // Possibly rename to `deny` for consistency
    sc->allow_shortcuts_inhibit = false;
} else if (strcmp(sc->name, "*") == 0) {
    return cmd_results(CMD_INVALID, "cannot activate/deactivate for wildcard seat");
} else if (!config->active) {
    return cmd_results(CMD_INVALID, "only enable and disable can be used in the config");
} else {
    // Validate seat exists
    // Validate inhibitor exists
    // Use parse_boolean(argv[0], inhibitor->active) to get state to set
}
```

**Comment by RedSoxFan** [comment]
I think `disable` should also deactivate all existing inhibitors on that seat. I don't think the inverse is true though. The proper place to deactivate them would be in https://github.com/swaywm/sway/blob/master/sway/input/seat.c#L1348

File: sway/commands/seat/shortcuts_inhibitor.c (hunk 7 above)
Comment (michaelweiser) [reply sway/commands/seat/shortcuts_inhibitor.c]: Done. Lost the operations enum. I stuck with enable/disable instead of allow/deny because deny in particular suggests that we have something to say in the matter, i.e. would actively refuse inhibitors. In think "disable" makes it more clear that we're just not going to honour the inhibitors. I pruned all other uses of "allow" to be consistent.

**Comment by michaelweiser** [comment]
> I think `disable` should also deactivate all existing inhibitors on that seat. I don't think the inverse is true though.

Agreed and Done.

> The proper place to deactivate them would be in https://github.com/swaywm/sway/blob/master/sway/input/seat.c#L1348

I did that but then, once inhibitors being disabled has become part of the config, it re-deactivates all inhibitors on any succeeding seat config command, including those that just activated one. I think it's a conflict between *changing the config* and *executing an action on a config change*. I agree that we should disable all active inhibitors once the user decides not to honour any new ones either but from then on the user should IMO also be able to explicitly (re-)activate them. So I made this part of the seat command for now, which somewhat complicates control flow in there. Alternatively I could think of somehow tracking changes through the config merge so that deactivation in `seat_apply_config()` only happens if the config item just changed *to* disabled from another value. Have I missed some existing infrastructure that permits that already?

File: sway/commands/shortcuts_inhibitor.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaelweiser) [reply sway/commands/shortcuts_inhibitor.c]: Done. I had totally missed `seat -`. I've gone the route of applying the active/inactive/toggle actions to the current/named/all seat(s) because it seems most consistent to me.

Is there a way to avoid iterating all seats and all inhibitors when trying to add an inhibitor indicator to `get_tree`? For now I've added a note to the manpage that indicator indication is up to the client.

**Comment by michaelweiser** [comment]
Is there anything I should be doing here?

File: sway/sway-input.5.scd [hunk 8]
```diff
@@ -239,6 +239,25 @@ correct seat.
 	by default) for the seat. This is primarily useful for video games. The
 	"escape" command can be used at runtime to escape from a captured client.
 
+*seat* <name> shortcuts_inhibitor enable|disable|active|inactive|toggle
```
Comment (emersion) [inline sway/sway-input.5.scd]: Maybe we should rename `active` to `activate`, just to make it clear the command performs an action and isn't read-only.

File: sway/commands/seat/shortcuts_inhibitor.c [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	if (error) {
+		return error;
+	}
+
+	struct seat_config *sc = config->handler_context.seat_config;
+	if (!sc) {
+		return cmd_results_new(CMD_FAILURE, "No seat defined");
+	}
+
+	if (strcmp(argv[0], "enable") == 0) {
```
Comment (emersion) [inline sway/commands/seat/shortcuts_inhibitor.c]: When used at runtime, this should probably also deactivate any active inhibitor

Comment (michaelweiser) [reply sway/commands/seat/shortcuts_inhibitor.c]: I had thought about *activating* all existing inhibitors when enabling the overall functionality but that seems overzealous to me. I agree with @RedSoxFan that disabling the functionality should also deactivate all active inhibitors for good measure and consistency but why should enabling it do the same?

Comment (emersion) [reply sway/commands/seat/shortcuts_inhibitor.c]: Oh, sorry, I misread your code. I agree, only `disable` should be a special-case.

**Comment by emersion** [comment]
Follow-up issue: decide how to let users allow/deny keyboard shortcut inhibition requests on a per-client basis.

* `for_window […] shortcuts_inhibitor enable|disable`?
* Somehow add a way to run a command which returns zero to allow the client to inhibit shortcuts?
* Leave this feature for the more general Wayland protocol security framework (https://github.com/swaywm/sway/issues/2333)?

File: sway/sway-input.5.scd (hunk 8 above)
Comment (michaelweiser) [reply sway/sway-input.5.scd]: @RedSoxFan suggested the use of `parse_boolean()` which handles the special `toggle` case and accepts `yes`, `on`, `true`, `enable`, `enabled`, and `active`. Neither of those matches particularly well and I also don't like the fact that it silently accepts anything else as false. So I would eliminate it again and explicitly match for `activate`, `deactivate` and `toggle` as before, perhaps without the enum boilerplate. Would that be acceptable? Alternatively, we could extend `parse_boolean()`...

File: sway/commands/seat/shortcuts_inhibitor.c (hunk 9 above)
Comment (michaelweiser) [reply sway/commands/seat/shortcuts_inhibitor.c]: Ah, well, if it's easily misread then maybe control flow is too complicated now. Do you have a suggestion where I maybe didn't fully understand @RedSoxFan's suggestion of doing the disabling int `seat_apply_config()`?

**Comment by michaelweiser** [comment]
> Follow-up issue: decide how to let users allow/deny keyboard shortcut inhibition requests on a per-client basis.
> 
>     * `for_window […] shortcuts_inhibitor enable|disable`?

Apart from issue of addressing the currently focused window with `seat -`, the following should allow inhibitors in general but immediately deactivate them for specific kinds of windows:
```
seat * shortcuts_inhibitor enable
for_window [...] seat - shortcuts_inhibitor deactivate
```
But since criteria seems to be applied only once, this would not guard against the client destroying and reestablishing a new inhibitor, which they seem to do quite aggressively.
So we'd need a per-window/view, non-per-seat version of the command and a per-window/view flag for it, right?

>     * Somehow add a way to run a command which returns zero to allow the client to inhibit shortcuts?
I don't get this one. Can you elaborate?

>     * Leave this feature for the more general Wayland protocol security framework (#2333)?
I'd rather do this now and port to a more generic framework later than to defer (indefinitely).

**Comment by emersion** [comment]
>So we'd need a per-window/view, non-per-seat version of the command and a per-window/view flag for it, right?

Yeah, that's what I meant. I think we should leave it for a follow-up PR though, to keep this one easier to review.

File: sway/sway-input.5.scd (hunk 8 above)
Comment (michaelweiser) [reply sway/sway-input.5.scd]: Switched from parse_boolean back to explicit parsing. Also simplified the control flow. I could comfortably live with this latest incarnation, fwiw.

**Comment by michaelweiser** [comment]
> I think we should leave it for a follow-up PR though, to keep this one easier to review.

Agreed.

File: sway/commands/seat/shortcuts_inhibitor.c [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			if (!seat) {
+				return cmd_results_new(CMD_FAILURE,
+						"Seat %s does not exist", sc->name);
+			}
+			error = handle_action(sc, seat, argv[0]);
+		} else {
+			struct sway_seat *seat = NULL;
+			wl_list_for_each(seat, &server.input->seats, link) {
+				error = handle_action(sc, seat, argv[0]);
+				if ((error && error->status != CMD_SUCCESS)) {
```
Comment (RedSoxFan) [inline sway/commands/seat/shortcuts_inhibitor.c]: only one set of parenthesis are needed

Comment (michaelweiser) [reply sway/commands/seat/shortcuts_inhibitor.c]: Removed, also in `cursor.c` where they came from by separate patch. Squashed the commits, quickly retested that everything still works as expected. Ready to go from my side.

**Comment by RedSoxFan** [comment]
Thanks!
